apigee_edge.module

```diff
@@ -1270,6 +1270,36 @@ function template_preprocess_app_credential(array &$variables) {
   }
 }
 
+/**
+ * Forcing user email address to lowercase for ApigeeX/Hybrid Org.
+ * @see https://github.com/apigee/apigee-edge-drupal/issues/730
+ */
+function apigee_edge_user_email_check(&$account) {
```

Comment:
Sorry, but I believe this should be a new[ field level constraint](https://www.drupal.org/docs/drupal-apis/entity-api/entity-validation-api/defining-constraints-validations-on-entities-andor-fields) which would run whenever and wherever `$user->validate()` or `$userl->mail->validate()` is called. This would be the standard Drupal/Symfony way or ensuring data consistency.

As a last result, a sanity check could still be executed in `apigee_edge_user_presave()` to ensure the above described field level validation was actually validated before entity save - because we all know that some IDP integration modules loves to bypass validating the user entity before they save it - and if an email still contains uppercase characters a `LogicException` (or a subclass of it) could be thrown to prevent user creation. (Which is still better than spending hours on figuring our why they is a Drupal user in the system without a developer account on the Apigee side.)

Reply:
I know of a case where the client's  LDP was using upper-case characters in every user's email address, which apparently was getting synced with Apigee. The client was reluctant to change the LDP configurations so wrote a hook to convert the new user's email address to lowercase and also an update hook to convert the existing user's email address to lowercase which in. turn syncs the developer account Apigee as well.